src/shared/sleep-config.c

```diff
@@ -435,5 +419,65 @@ static int can_sleep_internal(const char *verb, bool check_allowed) {
 }
 
 int can_sleep(const char *verb) {
-        return can_sleep_internal(verb, true);
+        int r;
+        _cleanup_(free_sleep_configp) SleepConfig *sleep_config = NULL;
+
+        r = parse_sleep_config(verb, &sleep_config);
+        if (r < 0)
+                return r;
+
+        return can_sleep_internal(verb, true, sleep_config);
+}
+
+int sleep_settings(const char *verb, SleepConfig *sleep_config, bool *ret_allow, char ***ret_modes, char ***ret_states) {
+
+        if (!verb || !sleep_config)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Invalid sleep configuration values provided");
+
+        assert(STR_IN_SET(verb, "suspend", "hibernate", "hybrid-sleep", "suspend-then-hibernate"));
+
+        if (streq(verb, "suspend")) {
+                *ret_allow = sleep_config->allow_suspend != 0;
+                *ret_modes = sleep_config->suspend_modes;
+                *ret_states = sleep_config->suspend_states;
+        } else if (streq(verb, "hibernate")) {
+                *ret_allow = sleep_config->allow_hibernate != 0;
+                *ret_modes = sleep_config->hibernate_modes;
+                *ret_states = sleep_config->hibernate_states;
+        } else if (streq(verb, "hybrid-sleep")) {
+                *ret_allow = sleep_config->allow_hybrid_sleep > 0 ||
+                        (sleep_config->allow_suspend != 0 && sleep_config->allow_hibernate != 0);
+                *ret_modes = sleep_config->hybrid_modes;
+                *ret_states = sleep_config->hybrid_states;
+        } else if (streq(verb, "suspend-then-hibernate")) {
+                *ret_allow = sleep_config->allow_s2h > 0 ||
+                        (sleep_config->allow_suspend != 0 && sleep_config->allow_hibernate != 0);
+                *ret_modes = *ret_states = NULL;
+        }
+
+        /* suspend modes empty by default */
+        if ((!ret_modes && !streq(verb, "suspend")) || !ret_states) {
+                *ret_allow = false;
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "No modes or states set for %s; Check sleep.conf", verb);
+        }
+
+        return 0;
+}
+
+void free_sleep_config(SleepConfig *sc) {
+        if (!sc)
+                return;
+
+        sc->suspend_modes = strv_free(sc->suspend_modes);
```

Comment:
Not that it matters too much, but since `sc` is going to be unconditionally freed just below, those assignments are not necessary. Just call `strv_free` and ignore the result.